daemons/fenced/fenced_history.c

```diff
@@ -98,6 +101,109 @@ stonith_fence_history_cleanup(xmlNode *request, const char *target)
     }
 }
 
+/* keeping the length of fence-history within bounds
+ * =================================================
+ *
+ * If things are really running wild a lot of fencing-attempts
+ * might fill up the hash-map, eventually using up a lot
+ * of memory and creating huge history-sync messages.
+ * Before the history being synced across nodes at least
+ * the reboot of a cluster-node helped keeping the
+ * history within bounds even though not in a reliable
+ * manner.
+ *
+ * stonith_remote_op_list isn't sorted for time-stamps
+ * thus it would be kind of expensive to delete e.g.
+ * the oldest entry if it would grow past MAX_STONITH_HISTORY
+ * entries.
+ * It is more efficient to purge MAX_STONITH_HISTORY/2
+ * entries whenever the list grows beyond MAX_STONITH_HISTORY.
+ * (sort for age + purge the MAX_STONITH_HISTORY/2 oldest)
+ * That done on a per-node-base might raise the
+ * probability of large syncs to occur.
+ * Things like introducing a broadcast to purge
+ * MAX_STONITH_HISTORY/2 entries or not sync above a certain
+ * threshold coming to mind ...
+ * Simplest thing though is to purge the full history
+ * throughout the cluster once MAX_STONITH_HISTORY is reached.
+ * On the other hand this leads to purging the history in
+ * situations where it would be handy to have it probably.
+ */
+
+
+static int
+op_time_sort(const void *a_voidp, const void *b_voidp)
+{
+    const remote_fencing_op_t **a = (const remote_fencing_op_t **) a_voidp;
+    const remote_fencing_op_t **b = (const remote_fencing_op_t **) b_voidp;
+    gboolean a_pending = ((*a)->state != st_failed) && ((*a)->state != st_done);
+    gboolean b_pending = ((*b)->state != st_failed) && ((*b)->state != st_done);
+
+    if (a_pending && b_pending) {
+        return 0;
+    } else if (a_pending) {
+        return -1;
+    } else if (b_pending) {
+        return 1;
+    } else if ((*b)->completed == (*a)->completed) {
+        return 0;
+    } else if ((*b)->completed > (*a)->completed) {
+        return 1;
+    }
+
+    return -1;
+}
+
+
+/*!
+ * \internal
+ * \brief Do a local history-trim to MAX_STONITH_HISTORY / 2 entries
+ *        once over MAX_STONITH_HISTORY
+ */
+void
+stonith_fence_history_trim(void)
+{
+    guint num_ops;
+
+    if (!stonith_remote_op_list) {
+        return;
+    }
+    num_ops = g_hash_table_size(stonith_remote_op_list);
+    if (num_ops > MAX_STONITH_HISTORY) {
+        remote_fencing_op_t *ops[num_ops];
+        remote_fencing_op_t *op = NULL;
+        GHashTableIter iter;
+        int i;
+
+        crm_trace("Fencing History growing beyond limit of %d so purge "
+                  "half of failed/successful attempts", MAX_STONITH_HISTORY);
+
+        /* write all ops into an array */
+        for (i=0, g_hash_table_iter_init(&iter, stonith_remote_op_list);
```

Comment:
I believe some static analyzers will complain about a void value in a comma statement -- better to put it separate before.

Reply:
I try to avoid 'for' with parts of the initialization before and parts contained.
So let's go with the common 'while' on hash-table iteration ...